### PR TITLE
Delete TAGMESSAGE after tag command completes

### LINKED ISSUE #4358: When creating an annotated tag, a TAGMESSAGE file is left in the repository
When creating an annotated tag, a TAGMESSAGE file is left in the repository afterwards.

- GitExtensions version: 2.51RC1
- GIT version: 2.15.1.2
- OS version: Windows 10 Pro version 1709

**Did this work in previous version of GitExtensions (which)?**
This did not happen in any previous version of GitExtensions that I know of. (such as 2.50.2)

### PR DESCRIPTION
**Pull request by glen-nicol**
Fixes #4358

Changes proposed in this pull request:
 - Move the UI and command execution into the GitTagController where it can then cleanup the temporary file.

What did I do to test the code and ensure quality:
 - Added 3 testcases for annotated tags.
 - Throw if the exceution function is null
 - Deletes the temporary tag regardless of the execution result.

Has been tested on (remove any that don't apply):
 - GIT 2.10 and above
 - Windows 10

### CODE REVIEW

File: GitCommands/Git/Tag/GitTagController.cs [hunk 1]
```diff
@@ -42,7 +51,17 @@ public GitCreateTagCmd GetCreateTagCommand(GitCreateTagArgs args)
...
+            {
+                return presentUI(createTagCmd);
+            }
+            finally
+            {
+                if(tagMessageFileName != null && _fileSystem.File.Exists(tagMessageFileName))
+                {
+                    _fileSystem.File.Delete(tagMessageFileName);
+                }
+            }
```
Comment (RussKie) [inline GitCommands/Git/Tag/GitTagController.cs]: It is a good start, but I think we can do better than this.
I don't think UI must be concerned with how a tag is created, all it must be concerned with - the tag is created or not.
So your initial proposal is in the right direction - abstract tag file handling in the controller, however I don't like the proposed API.

Can we perhaps do something like the following?

```
    public class GitTagController : IGitTagController
    {
        private readonly IGitUICommands _uiCommand;

        public GitTagController(Func<string> getWorkingDir, IGitUICommands uiCommand, IFileSystem fileSystem)
        {
            _uiCommand = uiCommand;
        }

        public bool CreateTag(Form owner, GitCreateTagArgs args)
        {
            // create tag operations

            var success = _uiCommand.StartCommandLineProcessDialog(createTagCmd, owner);

            // if success - delete the tag file

            return success;
        }
    }
```

Comment (glen-nicol) [reply GitCommands/Git/Tag/GitTagController.cs]: Yes I agree that the ui shouldn't need to have knowledge of that command object. I'll make those changes. Just to be sure though the temp file needs to be deleted regardless of success right?

Comment (RussKie) [reply GitCommands/Git/Tag/GitTagController.cs]: Yes, I suppose that is correct

Comment (RussKie) [reply GitCommands/Git/Tag/GitTagController.cs]: The method must be renamed as well to reflect its new purpose

Comment (glen-nicol) [reply GitCommands/Git/Tag/GitTagController.cs]: Doh. Ha that seems obvious!

**Comment by ideaalab** [comment]
While this is fixed, how do users proceed?
I have a TAGMESSAGE file in the root directory. Should i delete it? Move it? Commit?
Thanks!

**Comment by RussKie** [comment]
TAGMESSAGE file is generated by GE while creating a tag. We have regressed in 2.51 and not deleted it after tag would get created.
If you have this file hanging about - delete it.

This fix is going out in the next maintenance release.

**Pull request merged**